substrate/frame/collective/src/migrations/v5.rs

```diff
@@ -0,0 +1,196 @@
+impl<T: Config<I>, I: 'static> OnRuntimeUpgrade for VersionUncheckedMigrateToV5<T, I> {
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, sp_runtime::TryRuntimeError> {
+		log::info!("pre-migration collective v5");
+		let count = old::ProposalOf::<T, I>::iter().count();
+		if old::Proposals::<T, I>::get().len() != count {
+			log::info!("collective proposals count inconsistency");
+		}
+		if old::Members::<T, I>::get().len() > <T as Config<I>>::MaxMembers::get() as usize {
+			log::info!("collective members exceeds MaxMembers");
+		}
+
+		Ok((old::Proposals::<T, I>::get()).encode())
+	}
+
+	fn on_runtime_upgrade() -> Weight {
+		let mut weight = Weight::zero();
+
+		// ProposalOf
+		for (hash, proposal) in old::ProposalOf::<T, I>::drain() {
+			weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+			let Ok(new_hash) = <T as Config<I>>::Preimages::note(proposal.encode().into()) else {
+				log::info!(
+					target: "runtime::collective",
+					"Failed to note preimage for proposal {:?}",
+					hash,
+				);
+				continue
+			};
+			weight = weight.saturating_add(T::DbWeight::get().writes(1));
+			if new_hash != hash {
+				log::info!(
+					target: "runtime::collective",
+					"Preimage hash mismatch for proposal, expected {:?}, got {:?}",
+					hash,
+					new_hash,
+				);
+			}
+			if !<T as Config<I>>::Preimages::is_requested(&new_hash) {
+				log::info!(
+					target: "runtime::collective",
+					"Preimage for proposal {:?} was not requested",
+					hash,
+				);
+			}
+		}
+
+		// Proposals
+		old::Proposals::<T, I>::kill();
+		weight = weight.saturating_add(T::DbWeight::get().writes(1));
+
+		// Voting
+		for (hash, vote) in old::Voting::<T, I>::drain() {
+			weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 2));
+			crate::Voting::<T, I>::insert(
+				hash,
+				crate::Votes::<T::AccountId, BlockNumberFor<T>, <T as Config<I>>::MaxMembers> {
+					index: vote.index,
+					threshold: vote.threshold,
+					// the following operations are safe since the bound was previously enforced
+					// by runtime code
+					ayes: BoundedVec::truncate_from(vote.ayes),
+					nays: BoundedVec::truncate_from(vote.nays),
+					end: vote.end,
+				},
+			);
+		}
+
+		// Members
+		crate::Members::<T, I>::put(BoundedVec::truncate_from(old::Members::<T, I>::get()));
+		weight.saturating_add(T::DbWeight::get().reads_writes(1, 1))
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade(state: Vec<u8>) -> Result<(), sp_runtime::TryRuntimeError> {
+		use frame_support::ensure;
+		log::info!("post-migration collective v5");
+
+		let old_proposals = Vec::<T::Hash>::decode(&mut state.as_slice()).map_err(|_| {
+			"the state parameter should be something that was generated by pre_upgrade"
+		})?;
+
+		ensure!(
+			old_proposals.len() as u32 == crate::Voting::<T, I>::count(),
+			"the number of proposals should be the same",
+		);
+		for old in old_proposals {
```

Comment:
can we also check that the state of each new proposal is correct (relative to the old proposal state)?

Reply:
Proposal is saved in a key-value store, where the key is the hash of the value, and is being migrated to Preimages.

We are moving the content here

https://github.com/paritytech/polkadot-sdk/blob/7db06f139e23dd60536466530fe836c743cb3f31/substrate/frame/collective/src/migrations/v5.rs#L98

did you mean that I should do an integrity check on the content in the post-upgrade check as well?

Reply:
Yes, `post_upgrade` should check every piece of data migrated against the pre-upgrade data and ensure it's in the correct